### ISSUE TITLE
Segfault on dependency update from OCI registry

### ISSUE THREAD

**Issue by bdabelow** [helm/helm#9267]
I am getting a segfault when trying to pull dependencies from an OCI registry.

Output of `helm version`:
```
version.BuildInfo{Version:"v3.5.0", GitCommit:"32c22239423b3b4ba6706d450bd044baffdcf9e6", GitTreeState:"clean", GoVersion:"go1.15.6"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:28:09Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Setup:**
```
export HELM_EXPERIMENTAL_OCI=1

# Create chart test-dep and push to registry
helm create test-dep
helm registry login myregistry.com
helm chart save test-dep myregistry.com/helm/test-dep:latest
helm chart push myregistry.com/helm/test-dep:latest

# Create chart test, having test-dep as a dependency
helm create test
echo -e '\n\ndependencies:\n  - name: test-dep\n    version: "0.1.0"\n    repository: "oci://myregistry.com/helm/test-dep:latest"' >>test/Chart.yaml
```
**Pull dependencies:**
```
helm dependency update test
```

**Expected outcome:**
test-dep is successfully pulled.

**Actual outcome:**
```
Getting updates for unmanaged Helm repositories...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x15fba23]

goroutine 14 [running]:
helm.sh/helm/v3/internal/experimental/registry.(*Client).PullChart(0x0, 0xc0000a6360, 0xc0000a6360, 0x0, 0x0)
	/home/circleci/helm.sh/helm/internal/experimental/registry/client.go:156 +0x183
helm.sh/helm/v3/pkg/getter.(*OCIGetter).get(0xc00082efa0, 0xc0003e4380, 0x38, 0x20, 0x1972720, 0x1a73901)
	/home/circleci/helm.sh/helm/pkg/getter/ocigetter.go:52 +0xc5
helm.sh/helm/v3/pkg/getter.(*OCIGetter).Get(0xc00082efa0, 0xc0003e4380, 0x38, 0xc0000a62c0, 0x4, 0x4, 0xc000194c00, 0x60, 0x58)
	/home/circleci/helm.sh/helm/pkg/getter/ocigetter.go:36 +0x77
helm.sh/helm/v3/pkg/repo.(*ChartRepository).DownloadIndexFile(0xc0000a9ae0, 0x460a80, 0xc000424060, 0xc000531748, 0x405835)
	/home/circleci/helm.sh/helm/pkg/repo/chartrepo.go:125 +0x330
helm.sh/helm/v3/pkg/downloader.(*Manager).parallelRepoUpdate.func1(0xc000817320, 0xc0008bfd10, 0xc0000a9ae0)
	/home/circleci/helm.sh/helm/pkg/downloader/manager.go:655 +0x45
created by helm.sh/helm/v3/pkg/downloader.(*Manager).parallelRepoUpdate
	/home/circleci/helm.sh/helm/pkg/downloader/manager.go:654 +0x91
```

The actual test was performed against an Azure Container Registry. I can temporarily provide a registry for testing if required.

**Comment by jdolitsky** [comment]
While the segfault is certainly a bug, the dependencies are not correctly specified.

With the current design, the tag must match the chart version. Can you publish `myregistry.com/helm/test-dep:0.1.1` and use a dependencies section that looks similar to the following?
```
dependencies:
  - name: test-dep
    version: "0.1.0"
    repository: "oci://myregistry.com/helm"
```

Note that the `repository` does not contain the chart name (it is assumed)

**Comment by bdabelow** [comment]
@jdolitsky Works like a charm. Thanks for your help.

**Comment by jdolitsky** [comment]
For users having this issue, please see documentation on proper way to define oci dependencies: https://helm.sh/docs/topics/registries/#specifying-dependencies

Will leave this open as it relates to the crash error

**Comment by ssc327** [comment]
I am not able to get this to work.

In my chart I have the following:
```
dependencies:
- name: incubator-mylib-chart
  version: "0.1.0"
  repository: "oci://acrdev001.azurecr.io/helm"
```

On my registry I see the chart and its tag.

When I run ```helm dependency update``` I get the following:

```
helm dependency update
Getting updates for unmanaged Helm repositories...
...Unable to get an update from the "oci://acrdev001.azurecr.io/helm" chart repository:
        tag explicitly required
Hang tight while we grab the latest from your chart repositories...
...Unable to get an update from the "acrdev001" chart repository (https://acrdev001.azurecr.io/helm/v1/repo):
        failed to fetch https://acrdev001.azurecr.io/helm/v1/repo/index.yaml : 401 Unauthorized
```

I think that 
```
...Unable to get an update from the "acrdev001" chart repository (https://acrdev001.azurecr.io/helm/v1/repo):
        failed to fetch https://acrdev001.azurecr.io/helm/v1/repo/index.yaml : 401 Unauthorized
```
Can be ignored since this is trying to access a chart repository. However the first error
```
...Unable to get an update from the "oci://acrdev001.azurecr.io/helm" chart repository:
        tag explicitly required
```
seems to be related to the OCI registry.

What am I missing?

**Comment by bdabelow** [comment]
@ssc327 Did you actually push your chart to acrdev001.azurecr.io/helm/incubator-mylib-chart? If you pushed to acrdev001.azurecr.io/incubator-mylib-chart instead you need to drop the "helm" from the repository: line.

**Comment by ssc327** [comment]
@bdabelow yes, I did push to acrdev001.azurecr.io/helm/incubator-mylib-chart. 
After your comment I also tried pushing to acrdev001.azurecr.io/incubator-mylib-chart and updated the dependency as well but the result was the same.

**Comment by kirinnee** [comment]
Hello, I can confirm this happens on JetbrainsSpace OCI repository too. 

I can run `helm chart` `push`, `pull` and `export` and `save` just fine.  However, `helm dependency update` results in ` manifest does not contain at least 1 layer (total: 0)` and `helm install <name> oci://<ref>` results in seg faults. 

Seg fault: 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x15fba43]

goroutine 1 [running]:
helm.sh/helm/v3/internal/experimental/registry.(*Client).PullChart(0x0, 0xc000869160, 0xc000869160, 0x0, 0x0)
        /home/circleci/helm.sh/helm/internal/experimental/registry/client.go:156 +0x183
helm.sh/helm/v3/pkg/getter.(*OCIGetter).get(0xc000b5d0e0, 0xc000bf3500, 0x49, 0xc000d8f8c0, 0xc000bf3500, 0x49)
        /home/circleci/helm.sh/helm/pkg/getter/ocigetter.go:52 +0xc5
helm.sh/helm/v3/pkg/getter.(*OCIGetter).Get(0xc000b5d0e0, 0xc000bf3500, 0x49, 0xc00045a8a0, 0x4, 0x6, 0xc000b5d0e0, 0x0, 0x0)
        /home/circleci/helm.sh/helm/pkg/getter/ocigetter.go:36 +0x77
helm.sh/helm/v3/pkg/downloader.(*ChartDownloader).DownloadTo(0xc000d8fb70, 0x7fff289511fc, 0x49, 0x0, 0x0, 0xc0004232f0, 0x22, 0xc00073d460, 0x7f69137ab218, 0xffffffffffffffff, ...)
        /home/circleci/helm.sh/helm/pkg/downloader/chart_downloader.go:99 +0x13f
helm.sh/helm/v3/pkg/action.(*ChartPathOptions).LocateChart(0xc00061ec88, 0x7fff289511fc, 0x49, 0xc00009e0e0, 0x7fff289511f9, 0x2, 0x7fff289511fc, 0x49)
        /home/circleci/helm.sh/helm/pkg/action/install.go:675 +0x40e
main.runInstall(0xc00087d3b0, 0x2, 0x7, 0xc00061ec80, 0xc0007045a0, 0x1e32f40, 0xc00000e018, 0x1bf5416, 0x4, 0x5daaa0)
        /home/circleci/helm.sh/helm/cmd/helm/install.go:185 +0x172
main.newInstallCmd.func2(0xc00079d8c0, 0xc00087d3b0, 0x2, 0x7, 0x0, 0x0)
        /home/circleci/helm.sh/helm/cmd/helm/install.go:120 +0x85
github.com/spf13/cobra.(*Command).execute(0xc00079d8c0, 0xc00087d340, 0x7, 0x7, 0xc00079d8c0, 0xc00087d340)
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0xc000714000, 0xc00011ae00, 0x1, 0xc00091ff60)
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
main.main()
        /home/circleci/helm.sh/helm/cmd/helm/helm.go:80 +0x24d
```

Dependency update:
```
Getting updates for unmanaged Helm repositories...
...Unable to get an update from the "oci://atomicloud.registry.jetbrains.space/p/ops/helm-charts" chart repository:
        tag explicitly required
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "traefik" chart repository
...Successfully got an update from the "banzai" chart repository
Update Complete. ⎈Happy Helming!⎈
Saving 1 charts
Downloading mychart from repo oci://atomicloud.registry.jetbrains.space/p/ops/helm-charts
0.1.2: Pulling from atomicloud.registry.jetbrains.space/p/ops/helm-charts/mychart
Save error occurred:  could not download oci://atomicloud.registry.jetbrains.space/p/ops/helm-charts/mychart: manifest does not contain at least 1 layer (total: 0)
Deleting newly downloaded charts, restoring pre-update state
Error: could not download oci://atomicloud.registry.jetbrains.space/p/ops/helm-charts/mychart: manifest does not contain at least 1 layer (total: 0)
```

Chart.yml (for dependency update)
```
dependencies:
  - name: mychart
    version: "0.1.2"
    repository: "oci://atomicloud.registry.jetbrains.space/p/ops/helm-charts"
```

Full reference of OCI image
```
atomicloud.registry.jetbrains.space/p/ops/helm-charts/mychart:0.1.2
```

Helm Version
```
version.BuildInfo{Version:"v3.5.1", GitCommit:"32c22239423b3b4ba6706d450bd044baffdcf9e6", GitTreeState:"clean", GoVersion:"go1.15.7"}
```

**Comment by bacongobbler** [comment]
Should be fixed via #9409. Let us know if that isn’t the case. Closing.
